opal/datatype/opal_datatype_get_count.c

```diff
@@ -143,3 +141,54 @@ int32_t opal_datatype_set_element_count( const opal_datatype_t* datatype, size_t
     }
 }
 
+/**
+ * Compute the array of counts of the predefined datatypes contained in
+ * the datatype. We have no simple way to create this array, as we only
+ * sporadically need it (when we deal with heterogeneous environments or
+ * when we use get_element_count). Thus, we will pay the cost once per
+ * datatype, but we will only update this array if/when needed.
+ */
+int opal_datatype_compute_ptypes( opal_datatype_t* datatype )
```

Comment:
the ```datatype/getel``` test from the ```ibm``` test suite loops forever.

i got things working with the patch below.
```opal_datatype_get_element_count``` might have to be updated as well, though i could not write a program that causes the same endless loop

```patch
diff --git a/opal/datatype/opal_datatype_get_count.c b/opal/datatype/opal_datatype_get_count.c
index a860d5f..c7c87ea 100644
--- a/opal/datatype/opal_datatype_get_count.c
+++ b/opal/datatype/opal_datatype_get_count.c
@@ -169,10 +169,13 @@ int opal_datatype_compute_ptypes( opal_datatype_t* datatype )
     while( 1 ) {  /* loop forever the exit condition is on the last OPAL_DATATYPE_END_LOOP */
         if( OPAL_DATATYPE_END_LOOP == pElems[pos_desc].elem.common.type ) { /* end of the current loop */
             if( --(pStack->count) == 0 ) { /* end of loop */
-                stack_pos--; pStack--;
-                if( stack_pos == -1 ) return 0;  /* completed */
+                if( stack_pos == 0 ) return 0;  /* completed */
+                stack_pos--;
+                pStack--;
+                pos_desc++;
+            } else {
+                pos_desc = pStack->index + 1;
             }
-            pos_desc = pStack->index + 1;
             continue;
         }
         if( OPAL_DATATYPE_LOOP == pElems[pos_desc].elem.common.type ) {
```